### ISSUE TITLE
Opaque types are interchangeable

### ISSUE THREAD

**Issue by natesilva** [sindresorhus/type-fest#70]
Do you have any examples for `Opaque`? I thought `Opaque` types weren’t supposed to be interchangeable.

```typescript
import { Opaque } from 'type-fest';

type AccountNumber = Opaque<number>;
type AccountBalance = Opaque<number>;

function getBalance(accountNumber: AccountNumber): AccountBalance {
  return 101 as AccountBalance;
}

// pass an AccountBalance instead of an AccountNumber
const ab = 42 as AccountBalance;
console.log(getBalance(ab)); 
// expected: should not compile
// actual: gets balance for account 42
```

**Comment by natesilva** [comment]
The example shown in https://codemix.com/opaque-types-in-javascript/ works as expected. Their version of `Opaque` is defined as:

```typescript
type Opaque<K, T> = T & { __TYPE__: K };
```

With that definition they are not interchangeable. So I think this is a bug?

**Comment by sindresorhus** [comment]
// @resynth1943

**Comment by resynth1943** [comment]
> The example shown in https://codemix.com/opaque-types-in-javascript/ works as expected. Their version of `Opaque` is defined as:
> 
> ```ts
> type Opaque<K, T> = T & { __TYPE__: K };
> ```
> 
> With that definition they are not interchangeable. So I think this is a bug?

That's basically the same as the definition of the `Opaque` type in this library. To be precise, it's
```ts
export type Opaque<Type> = Type & {readonly __opaque__: unique symbol};
```

I think I know why TypeScript is doing this...

> ```ts
> import { Opaque } from 'type-fest';
> 
> type AccountNumber = Opaque<number>;
> type AccountBalance = Opaque<number>;
> 
> function getBalance(accountNumber: AccountNumber): AccountBalance {
>  return 101 as AccountBalance;
> }
> ```

Now, as you can see, `AccountNumber` is basically equal to `AccountBalance`... from TypeScript's viewpoint. They both have the same properties and are basically interchangeable.

After some minutes of exploration, I think I've found a solution:

```ts
type Opaque<Type, Marker extends string> = Type & Record<Marker, 'opaque'>;
type A = Opaque<number, 'A'>;
type B = Opaque<number, 'B'>;

function useA(a: A) { }

function getA(): A {
    return 2 as A;
}

const a: A = getA();
const b: B = getA(); // Compilation error.
useA(a);
```

In short, providing unique names to the `Opaque` type seems to be the only way to make an opaque type truly *unique* from the rest, preventing typing errors.

@sindresorhus How do you feel about this?

**Comment by resynth1943** [comment]
If everyone's happy with my proposal above, I'd be happy to make a Pull Request. Following this, perhaps making the second generic parameter default to `unknown` would allow backwards compatibility?

**Comment by natesilva** [comment]
I’m happy with that proposal, including defaulting the second parameter to `unknown`.

**Comment by resynth1943** [comment]
Well if everyone's happy, I'm happy to blow the cobwebs off VSCodium and PR this in. :heart:

**Comment by sindresorhus** [comment]
👍

**Comment by natesilva** [comment]
Thank you!
